pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,386 @@
+	if !cache.WaitForCacheSync(stopCh, c.traceflowListerSynced) {
+		klog.Errorf("Unable to sync caches for %s", controllerName)
+		return
+	}
+	klog.Infof("Caches are synced for %s", controllerName)
+
+	go c.ofClient.ReceivePacketIn(c, stopCh)
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addTraceflow(obj interface{}) {
+	tf := obj.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s ADD event", tf.Name)
+	c.enqueueTraceflow(tf)
+}
+
+func (c *Controller) updateTraceflow(oldObj, curObj interface{}) {
+	tf := curObj.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s UPDATE event", tf.Name)
+	c.enqueueTraceflow(tf)
+}
+
+func (c *Controller) deleteTraceflow(old interface{}) {
+	tf := old.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s DELETE event", tf.Name)
+	c.deallocateTag(tf)
+}
+
+// worker is a long-running function that will continually call the processTraceflowItem function
+// in order to read and process a message on the workqueue.
+func (c *Controller) worker() {
+	for c.processTraceflowItem() {
+	}
+}
+
+// processTraceflowItem processes an item in the "traceflow" work queue, by calling syncTraceflow
+// after casting the item to a string (Traceflow name). If syncTraceflow returns an error, this
+// function logs error. If syncTraceflow is successful, the Traceflow is removed from the queue
+// until we get notified of a new change. This function returns false if and only if the work queue
+// was shutdown (no more items will be processed).
+func (c *Controller) processTraceflowItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// We call Done here so the workqueue knows we have finished processing this item. We also
+	// must remember to call Forget if we do not want this work item being re-queued. For
+	// example, we do not call Forget if a transient error occurs, instead the item is put back
+	// on the workqueue and attempted again after a back-off period.
+	defer c.queue.Done(obj)
+
+	// We expect strings (Traceflow name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again.
+		c.queue.Forget(key)
+	} else {
+		// If error occurs we log error.
+		klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+	}
+	return true
+}
+
+// syncTraceflow gets Traceflow CRD by name, update cache and start syncing.
+func (c *Controller) syncTraceflow(traceflowName string) error {
```

Comment:
I do not meant this PR, but later we can consider to let controller compute which Node is the sender, and each Node watch the TF CRD with some filter to get and process only TF from the Node. The receiver can get the TF CRD when receiving a TF packet from OVS.
@tnqn @wenyingd

Reply:
This is a good idea. I also suggest to make the traceflow flow entries permanently to reduce the event which agent need to process, if no performance issue.

Reply:
Added TODO

Reply:
@jianjuns I think CRD doesn't support field selector yet: https://github.com/kubernetes/kubernetes/issues/53459.
But label selector might work according to https://github.com/kubernetes/kubernetes/issues/53459#issuecomment-528267120

Reply:
Ok. Let us think about how to optimize later.